src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,206 @@ public interface IFileWatchingToken
+            {
+                Contract.ThrowIfFalse(kind is Kind.UnwatchFiles);
+                _kind = kind;
+
+                _tokens = tokens;
+
+                // Other watching fields are not used for this kind
+                _directory = null!;
+                _filter = null;
+                _fileChangeFlags = 0;
+                _sink = null!;
+                _cookies = null!;
+                _token = null!;
+            }
+
+            private WatcherOperation(Kind kind, IFileWatchingToken token)
+            {
+                Contract.ThrowIfFalse(kind is Kind.UnwatchFile);
+                _kind = kind;
+
+                _token = token;
+
+                // Other watching fields are not used for this kind
+                _directory = null!;
+                _filter = null;
+                _fileChangeFlags = 0;
+                _sink = null!;
+                _cookies = null!;
+                _tokens = null!;
+            }
+
+            private enum Kind
+            {
+                None,
+                WatchDirectory,
+                WatchFile,
+                UnwatchFile,
+                UnwatchDirectories,
+                UnwatchFiles,
+            }
+
+            public static WatcherOperation Empty => new(Kind.None);
+
+            public static WatcherOperation WatchDirectory(string directory, string? filter, IVsFreeThreadedFileChangeEvents2 sink, List<uint> cookies)
+                => new(Kind.WatchDirectory, directory, filter, sink, cookies);
+
+            public static WatcherOperation WatchFile(string path, _VSFILECHANGEFLAGS fileChangeFlags, IVsFreeThreadedFileChangeEvents2 sink, IFileWatchingToken token)
+                => new(Kind.WatchFile, path, fileChangeFlags, sink, token);
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<IFileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(IFileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                combined = default;
+                return false;
+            }
+
+            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            {
+                switch (_kind)
+                {
+                    case Kind.None:
+                        return;
```

Comment:
since this is already a valuetask, i don't see why we need .TryCombine.   The batch loop can just call .ApplyAsync on a .None operation and it will immediatley break out without allocation cost.

Reply:
➡️ Covered by https://github.com/dotnet/roslyn/pull/55855#discussion_r695171416